### PR TITLE
pub with ext command

### PR DESCRIPTION
**Pull request by DoraALin**
Signed-off-by: 元守 <lulin@youzan.com>

### CODE REVIEW

File: internal/ext/ext.go [hunk 1]
```diff
@@ -6,20 +6,23 @@ import (
...
 const (
-	E_TAG_NOT_SUPPORT = "E_TAG_NOT_SUPPORT"
+	E_EXT_NOT_SUPPORT = "E_EXT_NOT_SUPPORT"
 	E_BAD_TAG	= "E_BAD_TAG"
+	E_INVALID_JSON_HEADER = "E_INVALID_JSON_HEADER"
 )
 
 var MAX_TAG_LEN = 100
-var validTagFmt = regexp.MustCompile(`^[a-zA-Z0-9]+$`)
+var validTagFmt = regexp.MustCompile(`^[a-zA-Z0-9_]+$`)
```
Comment (absolute8511) [inline internal/ext/ext.go]: `[a-zA-Z0-9_-]`

File: nsqd/channel.go [hunk 2]
```diff
@@ -1705,6 +1708,25 @@ exit:
...
+	var err error
+	switch msg.ExtVer {
+	case ext.TAG_EXT_VER:
+		msgTag = string(msg.ExtBytes)
+	case ext.JSON_HEADER_EXT_VER:
+		var jsonExt map[string]interface{}
+		err = json.Unmarshal(msg.ExtBytes, &jsonExt)
+		if err == nil {
+			if iMsgTag, exist := jsonExt[ext.CLEINT_DISPATCH_TAG_KEY]; exist {
+				msgTag, _ = iMsgTag.(string)
```
Comment (absolute8511) [inline nsqd/channel.go]: need check if ok here

File: nsqd/message.go [hunk 3]
```diff
@@ -323,7 +324,23 @@ func decodeMessage(b []byte, isExt bool) (*Message, error) {
 		extVer := ext.ExtVer(uint8(b[bodyStart]))
 		msg.ExtVer = extVer
 		switch extVer {
+		case ext.JSON_HEADER_EXT_VER:
+			msg.ExtVer = ext.JSON_HEADER_EXT_VER
```
Comment (absolute8511) [inline nsqd/message.go]: all the `msg.ExtBytes` can be the same

File: nsqdserver/http.go [hunk 4]
```diff
@@ -56,6 +56,7 @@ func newHTTPServer(ctx *context, tlsEnabled bool, tlsRequired bool) *httpServer
 
 	// v1 negotiate
 	router.Handle("POST", "/pub", http_api.Decorate(s.doPUB, http_api.NegotiateVersion))
+	router.Handle("POST", "/pub_with_ext", http_api.Decorate(s.doPUB, http_api.NegotiateVersion))
```
Comment (absolute8511) [inline nsqdserver/http.go]: use `pub_ext` and use a different method `s.doPUBExt`

File: nsqdserver/http.go [hunk 5]
```diff
@@ -231,13 +232,19 @@ func (s *httpServer) getExistingTopicFromQuery(req *http.Request) (url.Values, *
...
-	return s.internalPUB(w, req, ps, false)
+	var pubExt bool
+	if strings.HasSuffix(req.URL.Path, "/pub_with_ext") {
+		pubExt = true
+	}
+	return s.internalPUB(w, req, ps, false, pubExt)
 }
 
-func (s *httpServer) internalPUB(w http.ResponseWriter, req *http.Request, ps httprouter.Params, enableTrace bool) (interface{}, error) {
+var HTTP_EXT_HEADER_PREFIX = "X-Nsqext-"
```
Comment (absolute8511) [inline nsqdserver/http.go]: Should be const

File: nsqdserver/protocol_v2.go [hunk 6]
```diff
@@ -1381,16 +1396,24 @@ func internalPubAsync(clientTimer *time.Timer, msgBody *bytes.Buffer, topic *nsq
...
+}
+
 func (p *protocolV2) internalPubAndTrace(client *nsqd.ClientV2, params [][]byte, traceEnable bool) ([]byte, error) {
 	startPub := time.Now().UnixNano()
 	bodyLen, topic, extContent, err := p.preparePub(client, params, p.ctx.getOpts().MaxMsgSize, false)
 	if err != nil {
 		return nil, err
 	}
+
+	pubWExt := isPubWithExt(params[0])
```
Comment (absolute8511) [inline nsqdserver/protocol_v2.go]: Maybe use outside var to indicate the ext pub

File: nsqdserver/http.go [hunk 7]
```diff
@@ -279,30 +286,68 @@ func (s *httpServer) internalPUB(w http.ResponseWriter, req *http.Request, ps ht
 
 	if s.ctx.checkForMasterWrite(topic.GetTopicName(), topic.GetTopicPart()) {
 		var err error
-		traceIDStr := params.Get("trace_id")
```
Comment (absolute8511) [inline nsqdserver/http.go:282]: Why removing this?

File: nsqdserver/http.go [hunk 8]
```diff
@@ -279,30 +286,68 @@ func (s *httpServer) internalPUB(w http.ResponseWriter, req *http.Request, ps ht
...
+			}
+			var jsonHeaderExt map[string]interface{}
+			err = json.Unmarshal([]byte(headerStr), &jsonHeaderExt)
 			if err != nil {
-				nsqd.NsqLogger().Logf("parse tag err: %v", err)
-				return nil, http_api.Err{400, ext.E_BAD_TAG}
+				return nil, http_api.Err{400, ext.E_INVALID_JSON_HEADER}
+			}
+			//check trace id
+			traceIDI, exist := jsonHeaderExt[ext.TRACE_ID_KEY]
```
Comment (absolute8511) [inline nsqdserver/http.go]: get the `traceid` after the http header parse

File: internal/ext/ext.go [hunk 9]
```diff
@@ -50,24 +56,39 @@ func (tag TagExt) GetTagName() string {
...
+func ValidateTag(beValidated string) error {
+	if valid := _validateTag([]byte(beValidated)); !valid {
+		return fmt.Errorf("invalid tag %v", beValidated)
+	}
+	return nil
+}
+
 //pass in []byte not nil
-func validateTag(beValidated []byte) bool {
+func _validateTag(beValidated []byte) bool {
```
Comment (absolute8511) [inline internal/ext/ext.go]: I think `_validateTag ` can be removed

File: nsqdserver/protocol_v2.go [hunk 10]
```diff
@@ -1395,11 +1425,39 @@ func (p *protocolV2) internalPubAndTrace(client *nsqd.ClientV2, params [][]byte,
...
+	if traceEnable && !pubExt {
 		traceID = binary.BigEndian.Uint64(messageBody[:nsqd.MsgTraceIDLength])
 		realBody = messageBody[nsqd.MsgTraceIDLength:]
+	} else if pubExt {
+		//read two byte header length
+		extJsonLen = binary.BigEndian.Uint16(messageBody[:nsqd.MsgJsonHeaderLength])
+		extJsonBytes := messageBody[nsqd.MsgJsonHeaderLength:nsqd.MsgJsonHeaderLength + extJsonLen]
+		jhe, ok := extContent.(*ext.JsonHeaderExt)
+		if !ok {
+			return nil, fmt.Errorf("invalid ext content type, Json Header Ext expected, got %v", reflect.TypeOf(extContent))
```
Comment (absolute8511) [inline nsqdserver/protocol_v2.go]: `reflect.Type` can be removed since the %v print is enough

File: nsqdserver/protocol_v2.go [hunk 11]
```diff
@@ -1395,11 +1425,39 @@ func (p *protocolV2) internalPubAndTrace(client *nsqd.ClientV2, params [][]byte,
...
+		}
+		//validate json header passin
+		jsonHeader, err := simpleJson.NewJson(extJsonBytes)
+		if err != nil {
+			return nil, protocol.NewClientErr(err, ext.E_INVALID_JSON_HEADER, "fail to parse json header")
+		}
+
+		//parse traceID, if there is any
+		traceIDJson := jsonHeader.Get(ext.TRACE_ID_KEY)
```
Comment (absolute8511) [inline nsqdserver/protocol_v2.go]: duplicate code?

File: nsqdserver/protocol_v2.go [hunk 12]
```diff
@@ -1395,11 +1425,39 @@ func (p *protocolV2) internalPubAndTrace(client *nsqd.ClientV2, params [][]byte,
...
+		traceIDJson, existInJsonHeader = jsonHeader.CheckGet(ext.TRACE_ID_KEY)
+		if existInJsonHeader {
+			traceID, err = traceIDJson.Uint64()
```
Comment (absolute8511) [inline nsqdserver/protocol_v2.go]: Maybe add var `needTracedRsp` and should set to true if we found valid traceid here

File: nsqdserver/protocol_v2.go [hunk 13]
```diff
@@ -1426,10 +1484,11 @@ func (p *protocolV2) internalPubAndTrace(client *nsqd.ClientV2, params [][]byte,
 		topic.GetDetailStats().UpdatePubClientStats(client.String(), client.UserAgent, "tcp", 1, false)
 		cost := time.Now().UnixNano() - startPub
 		topic.GetDetailStats().UpdateTopicMsgStats(int64(len(realBody)), cost/1000)
-		if !traceEnable {
-			return okBytes, nil
+
+		if existInJsonHeader || (traceEnable && !existInJsonHeader) {
```
Comment (absolute8511) [inline nsqdserver/protocol_v2.go]: Maybe add var `needTracedRsp` to stand for whether we need return with `getTracedResponse`

File: nsqdserver/http.go [hunk 14]
```diff
@@ -28,6 +28,8 @@ import (
 	"github.com/julienschmidt/httprouter"
 )
 
+const HTTP_EXT_HEADER_PREFIX = "X-Nsqext-"
```
Comment (absolute8511) [inline nsqdserver/http.go:31]: `X-NSQEXT-` ?

File: nsqdserver/http.go (hunk 7 above)
Comment (DoraALin) [reply nsqdserver/http.go:282]: parse trace id logic moved to line#340

File: nsqdserver/protocol_v2.go (hunk 11 above)
Comment (DoraALin) [reply nsqdserver/protocol_v2.go]: yes, will be removed.

File: nsqdserver/http.go (hunk 7 above)
Comment (absolute8511) [reply nsqdserver/http.go:282]: Not exactly, trace_id should be get always,  `enableTrace` control the response body.

File: nsqdserver/protocol_v2.go (hunk 13 above)
Comment (DoraALin) [reply nsqdserver/protocol_v2.go]: bool var added

File: nsqdserver/http.go (hunk 14 above)
Comment (DoraALin) [reply nsqdserver/http.go:31]: will interpret key xxx in header key X-Nsqext-xxx into lowercase.

File: nsqdserver/http.go (hunk 7 above)
Comment (DoraALin) [reply nsqdserver/http.go:282]: I see, will apply needTraceRsp to decide whether write trace info in resp or not.

**Review by absolute8511 — APPROVED** [review]
LGTM

**Pull request merged**